src/Microsoft.AspNetCore.Mvc.Core/Internal/VirtualFileResultExecutor.cs

```diff
@@ -31,48 +32,87 @@ public VirtualFileResultExecutor(ILoggerFactory loggerFactory, IHostingEnvironme
 
         public Task ExecuteAsync(ActionContext context, VirtualFileResult result)
         {
-            SetHeadersAndLog(context, result);
-            return WriteFileAsync(context, result);
+            var fileInfo = GetFileInformation(result);
+            if (fileInfo.Exists)
+            {
+                var lastModified = result.LastModified == null ? fileInfo.LastModified : result.LastModified;
+                var (range, rangeLength, serveBody) = SetHeadersAndLog(
+                    context,
+                    result,
+                    fileInfo.Length,
+                    lastModified,
+                    result.EntityTag);
+                if (serveBody)
+                {
+                    return WriteFileAsync(context, result, fileInfo, range, rangeLength);
+                }
+            }
+            else
+            {
+                var (range, rangeLength, serveBody) = SetHeadersAndLog(context, result, null, result.LastModified, result.EntityTag);
+                if (serveBody)
+                {
+                    return WriteFileAsync(context, result, fileInfo, range, rangeLength);
+                }
+            }
+
+            return Task.CompletedTask;
         }
 
-        private async Task WriteFileAsync(ActionContext context, VirtualFileResult result)
+        private async Task WriteFileAsync(ActionContext context, VirtualFileResult result, IFileInfo fileInfo, RangeItemHeaderValue range, long rangeLength)
         {
             var response = context.HttpContext.Response;
-            var fileProvider = GetFileProvider(result);
-
-            var normalizedPath = result.FileName;
-            if (normalizedPath.StartsWith("~", StringComparison.Ordinal))
+            if (!fileInfo.Exists)
             {
-                normalizedPath = normalizedPath.Substring(1);
+                throw new FileNotFoundException(
+                    Resources.FormatFileResult_InvalidPath(result.FileName), result.FileName);
             }
-
-            var fileInfo = fileProvider.GetFileInfo(normalizedPath);
-            if (fileInfo.Exists)
+            else
             {
+                if (range != null && rangeLength == 0)
+                {
+                    return;
+                }
                 var physicalPath = fileInfo.PhysicalPath;
                 var sendFile = response.HttpContext.Features.Get<IHttpSendFileFeature>();
                 if (sendFile != null && !string.IsNullOrEmpty(physicalPath))
                 {
-                    await sendFile.SendFileAsync(
-                        physicalPath,
-                        offset: 0,
-                        count: null,
-                        cancellation: default(CancellationToken));
+                    if (range != null)
+                    {
+                        await sendFile.SendFileAsync(
+                            physicalPath,
+                            offset: (range.From.HasValue) ? range.From.Value : 0L,
```

Comment:
That won't work. range.From? returns a long and long does not have a definition for Value